pages/company/visualizations/gh_company_affiliation.py

```diff
@@ -0,0 +1,275 @@
+@callback(
+    Output(VIZ_ID, "figure"),
+    [
+        Input("repo-choices", "data"),
+        Input(f"{PAGE}-{paramter_2}-{VIZ_ID}", "value"),
+        Input(f"{PAGE}-{paramter_1}-{VIZ_ID}", "value"),
+        Input(f"{PAGE}-date-picker-range-{VIZ_ID}", "start_date"),
+        Input(f"{PAGE}-date-picker-range-{VIZ_ID}", "end_date"),
+    ],
+    background=True,
+    prevent_initial_call=True,
+)
+def gh_company_affiliation_graph(repolist, checks, num, start_date, end_date):
+
+    # wait for data to asynchronously download and become available.
+    cache = cm()
+    df = cache.grabm(func=cq, repos=repolist)
+    while df is None:
+        time.sleep(1.0)
+        df = cache.grabm(func=cq, repos=repolist)
+
+    start = time.perf_counter()
+    logging.debug(f"{VIZ_ID}- START")
+
+    # test if there is data
+    if df.empty:
+        logging.debug(f"{VIZ_ID} - NO DATA AVAILABLE")
+        return nodata_graph
+
+    # function for all data pre processing, COULD HAVE ADDITIONAL INPUTS AND OUTPUTS
+    df = process_data(df, checks, num, start_date, end_date)
+
+    fig = create_figure(df)
+
+    logging.debug(f"{VIZ_ID} - END - {time.perf_counter() - start}")
+    return fig
+
+
+def process_data(df: pd.DataFrame, checks, num, start_date, end_date):
+    """Implement your custom data-processing logic in this function.
+    The output of this function is the data you intend to create a visualization with,
+    requiring no further processing."""
+
+    # convert to datetime objects rather than strings
+    df["created"] = pd.to_datetime(df["created"], utc=True)
+
+    # order values chronologically by COLUMN_TO_SORT_BY date
+    df = df.sort_values(by="created", axis=0, ascending=True)
+
+    # filter values based on date picker
+    if start_date is not None:
+        df = df[df.created >= start_date]
+    if end_date is not None:
+        df = df[df.created <= end_date]
+
+    # intital count of same company name in github profile
+    result = df.cntrb_company.value_counts(dropna=False)
+
+    # reset format for df work
+    df = result.to_frame()
+    df["company_name"] = df.index
+    df = df.reset_index()
+    df["company_name"] = df["company_name"].astype(str)
+    df = df.rename(columns={"index": "orginal_name", "cntrb_company": "contribution_count"})
+
+    # applies fuzzy matching comparing all rows to each other
+    df["match"] = df.apply(lambda row: func(df, row["company_name"]), axis=1)
+
+    # changes company name to match other fuzzy matches
+    for x in range(0, len(df)):
+        matches = df.iloc[x]["match"]
+        for y in matches:
+            df.loc[y, "company_name"] = df.iloc[x]["company_name"]
+            df.loc[y, "match"] = ""
+
+    # groups all same name company affiliation and sums the contributions
+    df = (
+        df.groupby(by="company_name")["contribution_count"]
+        .sum()
+        .reset_index()
+        .sort_values(by=["contribution_count"])
+        .reset_index(drop=True)
+    )
+
```

Comment:
above may need a refactor at some point but I see what you're doing and it works.